File: examples/php/ai_obs.php
<?php

declare(strict_types=1);

/**
 * AI Observability Platform — PHP client
 * Requires PHP >= 8.1, ext-curl, ext-json (both enabled by default)
 */
final class AiObs
{
    private static string  $ingestUrl    = '';
    private static string  $token        = '';
    private static ?string $currentRunId = null;
    private static ?string $currentAgent = null;
    private static array   $spanStack    = [];

    public static function configure(string $ingestUrl, string $token): void
    {
        self::$ingestUrl = rtrim($ingestUrl, '/');
        self::$token     = $token;
    }

    public static function run(string $agent, array $input, callable $fn): mixed
    {
        $runId   = self::uuid4();
        $spanId  = self::randSpan();
        $started = self::now();

        self::$currentRunId = $runId;
        self::$currentAgent = $agent;
        self::$spanStack    = [$spanId];

        self::emit([
            'run_id'     => $runId,
            'span_id'    => $spanId,
            'type'       => 'run.start',
            'agent'      => $agent,
            'started_at' => $started,
            'payload'    => ['agent' => $agent, 'input_hash' => self::hashOf($input)],
        ]);

        $ctx    = new RunContext($runId, $agent, $spanId);
        $status = 'succeeded';
        try {
            return $fn($ctx);
        } catch (\Throwable $e) {
            $status = 'failed';
            throw $e;
        } finally {
            self::emit([
                'run_id'     => $runId,
                'span_id'    => self::randSpan(),
                'type'       => 'run.end',
                'agent'      => $agent,
                'started_at' => $started,
                'ended_at'   => self::now(),
                'payload'    => [
                    'status'         => $status,
                    'total_steps'    => $ctx->steps,
                    'total_tokens'   => $ctx->totalTokens,
                    'total_cost_usd' => $ctx->totalCost,
                ],
            ]);
            self::$currentRunId = null;
            self::$currentAgent = null;
            self::$spanStack    = [];
        }
    }

    public static function observe(
        ?string $llm   = null,
        ?string $tool  = null,
        ?string $agent = null,
        callable $fn   = null
    ): mixed {
        $kind   = $llm ? 'llm' : ($tool ? 'tool' : ($agent ? 'agent' : null));
        $target = $llm ?? $tool ?? $agent;
        if (!$kind || !$fn) {
            throw new \InvalidArgumentException('observe requires exactly one of llm, tool, agent AND a callable');
        }
        if (!self::$currentRunId) {
            throw new \RuntimeException('observe must be called inside run()');
        }

        $parent   = end(self::$spanStack) ?: null;
        $spanId   = self::randSpan();
        $started  = self::now();
        $typeMap  = ['llm' => 'llm.call', 'tool' => 'tool.invoke', 'agent' => 'step.start'];
        $type     = $typeMap[$kind];
        $curAgent = self::$currentAgent;
        $runId    = self::$currentRunId;

        $error = null;
        try {
            return $fn();
        } catch (\Throwable $e) {
            $error = $e;
            throw $e;
        } finally {
            $payload = match ($kind) {
                'llm'   => ['model' => $target, 'finish_reason' => $error ? 'error' : 'stop'],
                'tool'  => ['tool' => $target, 'args_hash' => 'sha256:' . str_repeat('0', 64), 'status' => $error ? 'error' : 'ok'],
                'agent' => ['step' => 0, 'status' => $error ? 'error' : 'ok'],
            };
            self::emit(array_filter([
                'run_id'         => $runId,
                'span_id'        => $spanId,
                'parent_span_id' => $parent,
                'type'           => $type,
                'agent'          => $curAgent,
                'llm_model'      => $kind === 'llm' ? $target : null,
                'started_at'     => $started,
                'ended_at'       => self::now(),
                'payload'        => $payload,
            ], static fn($v) => $v !== null));
        }
    }

    /** Generates a signed token. Normally done server-side — for reference/testing only. */
    public static function issueToken(
        string $secret,
        string $orgId,
        string $name    = 'default',
        array  $scopes  = ['ingest.write'],
        int    $ttl     = 31_536_000
    ): string {
        $raw = json_encode(['org_id' => $orgId, 'scopes' => $scopes, 'exp' => time() + $ttl, 'name' => $name]);
        $b64 = rtrim(strtr(base64_encode($raw), '+/', '-_'), '=');
        $sig = substr(hash_hmac('sha256', $b64, $secret), 0, 32);
        return "ai_obs_v1.{$b64}.{$sig}";
    }

    public static function emit(array $event): void
    {
        $url  = self::$ingestUrl . '/v1/events';
        $body = json_encode([$event]);
        $ch   = curl_init($url);
        curl_setopt_array($ch, [
            CURLOPT_POST           => true,
            CURLOPT_POSTFIELDS     => $body,
            CURLOPT_HTTPHEADER     => [
                'Content-Type: application/json',
                'Authorization: Bearer ' . self::$token,
            ],
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_TIMEOUT        => 5,
        ]);
        curl_exec($ch);
        curl_close($ch);
    }

    private static function uuid4(): string
    {
        $b     = random_bytes(16);
        $b[6]  = chr(ord($b[6]) & 0x0f | 0x40);
        $b[8]  = chr(ord($b[8]) & 0x3f | 0x80);
        return vsprintf('%s%s-%s-%s-%s-%s%s%s', str_split(bin2hex($b), 4));
    }

    private static function randSpan(): string { return bin2hex(random_bytes(8)); }

    private static function now(): string
    {
        return (new \DateTime('now', new \DateTimeZone('UTC')))->format(\DateTime::ATOM);
    }

    private static function hashOf(mixed $v): string
    {
        return 'sha256:' . hash('sha256', json_encode($v));
    }
}

final class RunContext
{
    public int   $steps       = 0;
    public int   $totalTokens = 0;
    public float $totalCost   = 0.0;

    public function __construct(
        public readonly string $runId,
        public readonly string $agent,
        public readonly string $spanId,
    ) {}

    public function checkpoint(int $step, array $state): void
    {
        $this->steps++;
        AiObs::emit([
            'run_id'     => $this->runId,
            'span_id'    => bin2hex(random_bytes(8)),
            'type'       => 'checkpoint',
            'agent'      => $this->agent,
            'started_at' => (new \DateTime('now', new \DateTimeZone('UTC')))->format(\DateTime::ATOM),
            'payload'    => ['step' => $step, 'state_hash' => 'sha256:' . hash('sha256', json_encode($state))],
        ]);
    }

    public function handoff(string $to, string $reason = 'delegation', array $payload = []): void
    {
        AiObs::emit([
            'run_id'     => $this->runId,
            'span_id'    => bin2hex(random_bytes(8)),
            'type'       => 'handoff',
            'agent'      => $this->agent,
            'started_at' => (new \DateTime('now', new \DateTimeZone('UTC')))->format(\DateTime::ATOM),
            'payload'    => [
                'from'         => $this->agent,
                'to'           => $to,
                'reason'       => $reason,
                'payload_hash' => 'sha256:' . hash('sha256', json_encode($payload)),
            ],
        ]);
    }
}

File: examples/go/ai_obs.go
// Package aiobs is a minimal client for the AI Observability Platform.
// Zero external dependencies — stdlib only.
package aiobs

import (
	"bytes"
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"net/http"
	"strings"
	"sync"
	"time"
)

// --- configuration ---

var cfg struct {
	mu        sync.RWMutex
	ingestURL string
	token     string
}

// Configure sets the ingest URL and service token. Call once at startup.
func Configure(ingestURL, token string) {
	cfg.mu.Lock()
	defer cfg.mu.Unlock()
	cfg.ingestURL = strings.TrimRight(ingestURL, "/")
	cfg.token = token
}

// --- run context ---

type runState struct {
	runID     string
	agent     string
	spanStack []string
}

var state struct {
	mu      sync.Mutex
	current *runState
}

// RunContext carries per-run helpers passed to the Run callback.
type RunContext struct {
	RunID       string
	Agent       string
	Steps       int
	TotalTokens int
	TotalCost   float64
	spanID      string
}

// Checkpoint emits a checkpoint event and increments the step counter.
func (rc *RunContext) Checkpoint(step int, stateObj any) {
	rc.Steps++
	emit(event{ //nolint:errcheck
		RunID: rc.RunID, SpanID: randSpan(), Type: "checkpoint",
		Agent: rc.Agent, StartedAt: iso(time.Now().UTC()),
		Payload: map[string]any{"step": step, "state_hash": hashOf(stateObj)},
	})
}

// Handoff emits a handoff event.
func (rc *RunContext) Handoff(to, reason string, payload any) {
	if reason == "" {
		reason = "delegation"
	}
	emit(event{ //nolint:errcheck
		RunID: rc.RunID, SpanID: randSpan(), Type: "handoff",
		Agent: rc.Agent, StartedAt: iso(time.Now().UTC()),
		Payload: map[string]any{
			"from": rc.Agent, "to": to,
			"reason": reason, "payload_hash": hashOf(payload),
		},
	})
}

// --- Run ---

// Run emits run.start, invokes fn, then emits run.end.
func Run(agent string, input any, fn func(*RunContext) error) error {
	runID  := mustUUID()
	spanID := randSpan()
	started := time.Now().UTC()

	state.mu.Lock()
	state.current = &runState{runID: runID, agent: agent, spanStack: []string{spanID}}
	state.mu.Unlock()

	emit(event{ //nolint:errcheck
		RunID: runID, SpanID: spanID, Type: "run.start",
		Agent: agent, StartedAt: iso(started),
		Payload: map[string]any{"agent": agent, "input_hash": hashOf(input)},
	})

	rc := &RunContext{RunID: runID, Agent: agent, spanID: spanID}
	status := "succeeded"
	err := fn(rc)
	if err != nil {
		status = "failed"
	}

	emit(event{ //nolint:errcheck
		RunID: runID, SpanID: randSpan(), Type: "run.end",
		Agent: agent, StartedAt: iso(started), EndedAt: iso(time.Now().UTC()),
		Payload: map[string]any{
			"status":         status,
			"total_steps":    rc.Steps,
			"total_tokens":   rc.TotalTokens,
			"total_cost_usd": rc.TotalCost,
		},
	})

	state.mu.Lock()
	state.current = nil
	state.mu.Unlock()

	return err
}

// --- Observe ---

// Observe wraps fn with a telemetry event.
// Pass exactly one non-empty string in llm, tool, or agentName.
func Observe(llm, tool, agentName string, fn func() (any, error)) (any, error) {
	kind, target := "llm", llm
	switch {
	case tool != "":
		kind, target = "tool", tool
	case agentName != "":
		kind, target = "agent", agentName
	}

	state.mu.Lock()
	st := state.current
	state.mu.Unlock()
	if st == nil {
		return nil, fmt.Errorf("Observe must be called inside Run")
	}

	parent  := st.spanStack[len(st.spanStack)-1]
	spanID  := randSpan()
	started := time.Now().UTC()
	typeMap := map[string]string{"llm": "llm.call", "tool": "tool.invoke", "agent": "step.start"}
	curAgent := st.agent
	runID    := st.runID

	result, err := fn()

	llmModel := ""
	if kind == "llm" {
		llmModel = target
	}
	payload := buildPayload(kind, target, err)

	emit(event{ //nolint:errcheck
		RunID: runID, SpanID: spanID, ParentSpanID: parent,
		Type: typeMap[kind], Agent: curAgent, LLMModel: llmModel,
		StartedAt: iso(started), EndedAt: iso(time.Now().UTC()),
		Payload: payload,
	})

	return result, err
}

// IssueToken generates a signed ingest token.
// Normally done server-side via the ingest-api — for reference/testing only.
func IssueToken(secret, orgID, name string, scopes []string, ttlSeconds int64) string {
	p, _ := json.Marshal(map[string]any{
		"org_id": orgID, "scopes": scopes,
		"exp": time.Now().Unix() + ttlSeconds, "name": name,
	})
	b64 := base64.RawURLEncoding.EncodeToString(p)
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(b64))
	sig := hex.EncodeToString(mac.Sum(nil))[:32]
	return "ai_obs_v1." + b64 + "." + sig
}

// --- internal ---

type event struct {
	RunID        string         `json:"run_id"`
	SpanID       string         `json:"span_id"`
	ParentSpanID string         `json:"parent_span_id,omitempty"`
	Type         string         `json:"type"`
	Agent        string         `json:"agent,omitempty"`
	LLMModel     string         `json:"llm_model,omitempty"`
	StartedAt    string         `json:"started_at"`
	EndedAt      string         `json:"ended_at,omitempty"`
	Payload      map[string]any `json:"payload"`
}

func emit(e event) error {
	cfg.mu.RLock()
	url := cfg.ingestURL + "/v1/events"
	tok := cfg.token
	cfg.mu.RUnlock()

	body, _ := json.Marshal([]event{e})
	req, err := http.NewRequest(http.MethodPost, url, bytes.NewReader(body))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+tok)

	go func() { http.DefaultClient.Do(req) }() //nolint:errcheck
	return nil
}

func buildPayload(kind, target string, err error) map[string]any {
	status := "ok"
	if err != nil {
		status = "error"
	}
	switch kind {
	case "llm":
		fr := "stop"
		if err != nil {
			fr = "error"
		}
		return map[string]any{"model": target, "finish_reason": fr}
	case "tool":
		return map[string]any{"tool": target, "args_hash": "sha256:" + strings.Repeat("0", 64), "status": status}
	default:
		return map[string]any{"step": 0, "status": status}
	}
}

func randSpan() string {
	b := make([]byte, 8)
	rand.Read(b) //nolint:errcheck
	return hex.EncodeToString(b)
}

func mustUUID() string {
	b := make([]byte, 16)
	rand.Read(b) //nolint:errcheck
	b[6] = (b[6] & 0x0f) | 0x40
	b[8] = (b[8] & 0x3f) | 0x80
	return fmt.Sprintf("%08x-%04x-%04x-%04x-%012x",
		b[0:4], b[4:6], b[6:8], b[8:10], b[10:16])
}

func iso(t time.Time) string { return t.Format(time.RFC3339) }

func hashOf(v any) string {
	b, _ := json.Marshal(v)
	h := sha256.Sum256(b)
	return "sha256:" + hex.EncodeToString(h[:])
}
